File: binary_search/FindMinRotatedSortedArray.java
package binary_search;

public class FindMinRotatedSortedArray {
    
    // Problem: Given the sorted rotated array nums of unique elements, return the minimum element of this array.
    // You must write an algorithm that runs in O(log n) time.

    public int findMin(int[] nums) {
        int left = 0;
        int right = nums.length - 1;
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            
            // If mid element is greater than right element, 
            // then the minimum is in the right half
            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } 
            // If mid element is less than or equal to right element,
            // then the minimum is in the left half (including mid)
            else {
                right = mid;
            }
        }
        
        // When left == right, we found the minimum
        return nums[left];
    }

    public static void main(String[] args) {
        FindMinRotatedSortedArray solution = new FindMinRotatedSortedArray();
        
        // Test Case 1
        int[] nums1 = {3, 4, 5, 1, 2};
        System.out.println("Test Case 1: " + solution.findMin(nums1)); // Expected: 1

        // Test Case 2
        int[] nums2 = {4, 5, 6, 7, 0, 1, 2};
        System.out.println("Test Case 2: " + solution.findMin(nums2)); // Expected: 0
        
        // Test Case 3
        int[] nums3 = {11, 13, 15, 17};
        System.out.println("Test Case 3: " + solution.findMin(nums3)); // Expected: 11
    }
}

File: binary_search/find_min_rotated_sorted_array.py
from typing import List

class Solution:
    """
    Problem: Given the sorted rotated array nums of unique elements, return the minimum element of this array.
    You must write an algorithm that runs in O(log n) time.
    """
    def findMin(self, nums: List[int]) -> int:
        # TODO: Implement Binary Search Solution
        return 0

if __name__ == "__main__":
    solution = Solution()
    
    # Test Case 1
    nums1 = [3, 4, 5, 1, 2]
    print(f"Test Case 1: {solution.findMin(nums1)}") # Expected: 1

    # Test Case 2
    nums2 = [4, 5, 6, 7, 0, 1, 2]
    print(f"Test Case 2: {solution.findMin(nums2)}") # Expected: 0

    # Test Case 3
    nums3 = [11, 13, 15, 17]
    print(f"Test Case 3: {solution.findMin(nums3)}") # Expected: 11

